### ISSUE TITLE
File locking under DOSBox-X?

### ISSUE THREAD

**Issue by emendelson** [joncampbell123/dosbox-x#2134]
I'm trying to help out a user who works with the DataPerfect database application in a networked environment. He's been using vDos because it fully supports file locking to prevent corruption of the database, but vDos can't work smoothly with an associated program called dpMouse, so he is considering DOSBox-X instead. 

Question: does DOSBox-X fully support file locking, and, if so, has this feature been fully tested in a networked environment? I don't want to recommend anything that will endanger this user's data, and will be very grateful for any information.

**Comment by rderooy** [comment]
@emendelson some changes where made very recently to prevent dosbox-x from corrupting disk images when opened by multiple instances of dosbox-x. See https://github.com/joncampbell123/dosbox-x/issues/2069

But I don't know if this prevents corruption of regular files.

**Comment by emendelson** [comment]
@rderooy Thank you; the disk image locking doesn't apply to database and other files, I think. But I could be mistaken!

**Comment by rderooy** [comment]
I did a quick test, where I mounted a host directory in dosbox-x and used the bundled Edit application to create a text file, I then modified the file on the host (with edit still open in dosbox-x) and got no errors that the file was locked when I saved it.

But if I recall correctly, this is normal for DOS applications as they have no concept of sharing (and locking) files, unless you have an application using SHARE.EXE to prevent corruption. And the SHARE implementation in DOSBox-X is broken. It is really a fake-share just to appease programs looking for SHARE being loaded.

**Comment by emendelson** [comment]
The issue here is what the vDos web site describes as: "Network support with (verified) record locking, required by multi-user DOS applications to function reliably." This affects dBase, DataPerfect, and any other database program that can be accessed from multiple systems. xBase and DataPerfect support record-locking and file-locking internally; they don't use Share. 

As far as I know, this feature has never been implemented in DOSBox-X, but I'm asking here in case I missed something that got added recently.

**Comment by emendelson** [comment]
An expert authority on this question confirms that DOSBox-X does NOT support the kind of locking required for multi-user database operations. All such applications require vDos, and nothing else is safe to use. I'll warn the user whom I mentioned in the first post.

**Comment by Wengier** [comment]
@emendelson I had checked the DOSBox-X code previously to confirm that record-locking has been implemented for DOSBox-X for a long time, but *only on the Windows platform*, similar to vDos/Plus. I was in contact with a DOS database user previously who confirmed by himself that the database was correctly locked for his purpose.

@rderooy Record locking is not the same as creating a text file and then modifying the file on the host at the same time. Recording locking is a DOS API that is only activated if it has been called by a (usually database) application.

**Comment by emendelson** [comment]
@Wengier My expert authority may not have looked at DOSBox-X for a while. I've asked him to check it again. If the report is wrong, of course I'll delete what I wrote.

**Comment by Wengier** [comment]
@emendelson Please check out the updated code in #2135; the file locking may work on all platforms. For now please set “nocachedir=true” for best result.

**Comment by emendelson** [comment]
@Wengier I have no way of testing this, because I don't have a networked environment and a database. Since I have no direct experience, I'll drop out of this discussion. I remembered that this link provides further information in posts from you and Jos Schaars:

https://vdos.proboards.com/thread/317/ctrl-break

Again, I have absolutely NO direct knowledge of any of this.

**Comment by Wengier** [comment]
@emendelson With the new code “nocachedir=true” does further things than that. DOSBox-X is being updated constantly with changes, and I am sure Jos was testing older code.

**Comment by StuartClipper** [comment]
I have been testing DOSBOX-X running in a multi user environment alongside Hyper-V virtual machines with Windows 10 32 bit running NTVDM cmd.exe or command.com ... our application is written in Clipper and linked with Blinker, it's in daily use by several companies with between 3 and 10 concurrent users. The record and file locking is working very well. 

However, opening dbf files on mapped drives is much slower in DOSBOX-X then NTVDM. 

Opening the files on the network location: 
Hyper-V ~ 8 seconds
DOSBOX-X ~ 4 minutes 17 seconds
 
Compared to opening the same set of files on a local drive: 
Hyper-V ~ 6 seconds
DOSBOX-X ~ 17 seconds

Interestingy once the files are open the operational speed of the program is almost identical.

**Comment by Wengier** [comment]
@StuartClipper What setting do you use for the benchmark? For example, did you set ```nocachedir=true``` or ``` “nocachedir=false”? Also, do you have any test files? Thanks.

**Comment by StuartClipper** [comment]
My tests were all conducted with: nocachedir = true
The test files I used were a copy of a client's live data consisting of 38 x .DBF, 12 x .DBT and 69 x .NTX files, a total of 1.49GB. 
I am unable to provide a copy of these test files due to client confidentiality; however, I'll happily spend time anonymising a copy if that will help?

**Comment by scot-x** [comment]
@Wengier Have there been any further developments or tests in this area with reference to database applications? Currently supporting a customer who needs to move a dbase IV based application to a windows server 2022 environment. We are starting comparison tests with other emulators. 
The setup has 5-10 users logging in per RDP to their accounts and starting their copy of the dbase application but all share the same directory where the DBF files are. Meaning that there is no network in between because all are on the same machine.
Would like to have more confidence in the required file-locking capabilities.
We are working with the `nocachedir=true` setting

28 x DBF files and associated MDX files (a total of around 300-400MB)

**Comment by emendelson** [comment]
Have you tried vDos (www.vdos.info)? It was written from the start for database applications that require locking.

**Comment by scot-x** [comment]
> Have you tried vDos ([www.vdos.info](http://www.vdos.info))? It was written from the start for database applications that require locking.

Yes, we tried it but ran into problems. It looked promising at first. There were though strange side effects when putting the current recno of a table into a variable when the recno was either equal to 33001 or greater than 33001. Meaning we always had to put the recno through the round() function to ensure that we got an integer back for storage in the variable. There was another problem as well, which I can't remember it at the moment.

**Comment by rderooy** [comment]
Regarding vDos, you may want to check: https://www.vogons.org/viewtopic.php?f=32&t=92472
